Add initial symbol binding for function pointer types

333fred opened the pull request:
This adds `FunctionPointerTypeSymbol`, which represents a function pointer type. This type represents the actual signature of a function pointer by using a `MethodSymbol` of kind `FunctionPointerSignature`. This will hopefully simplify future steps around checking conversions. This PR does not implement any of these conversion steps currently, that will be in a follow up.

@dotnet/roslyn-compiler @AlekseyTs for review.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4:
```diff
@@ -337,11 +337,7 @@ array_rank_specifier
-
-modified_type
-  : modifier* type
+  : 'delegate' '*' syntax_token? '<' parameter (',' parameter)* '>'
```
you ended up going with parameter?

333fred replied:
> you ended up going with parameter?

I did, in order to share parameter binding code.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal class FunctionPointerTypeSymbol : TypeSymbol
```
doesn't implement a corresponding IFunctionPointerTypeSymbol?

333fred replied:
> doesn't implement a corresponding IFunctionPointerTypeSymbol?

No, I'm leaving the public API for a future follow-up PR.

AlekseyTs replied:
> doesn't implement a corresponding IFunctionPointerTypeSymbol?

In addition to the other response, internal C# symbols no longer implement public symbols interfaces directly.

---
In reply to: [358685039](https://github.com/dotnet/roslyn/pull/40431#discussion_r358685039) [](ancestors = 358685039)

RikkiGibson commented on src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:
```diff
@@ -365,7 +368,8 @@ public virtual ImmutableArray<CSharpAttributeData> GetReturnTypeAttributes()
         /// </summary>
-        public abstract Symbol AssociatedSymbol { get; }
+        public abstract Symbol? AssociatedSymbol { get; }
+#nullable disable
```
`#nullable restore`

RikkiGibson commented on src/Compilers/CSharp/Portable/Syntax/Syntax.xml:
```diff
@@ -248,9 +248,9 @@
-    <Field Name="Arguments" Type="SeparatedSyntaxList&lt;ModifiedType&gt;" MinCount="1">
+    <Field Name="Parameters" Type="SeparatedSyntaxList&lt;ParameterSyntax&gt;" MinCount="1">
       <PropertyComment>
-        <summary>SeparatedSyntaxList of ModifiedTypes representing the list of parameters and return type.</summary>
```
Was this changed because ModifiedType presented too many barriers to code reuse in practice?

333fred replied:
Precisely.

---
In reply to: [359004607](https://github.com/dotnet/roslyn/pull/40431#discussion_r359004607) [](ancestors = 359004607)

RikkiGibson commented on src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs:
```diff
@@ -1151,6 +1151,7 @@ private BlobHandle GetMethodSignatureHandleAndBlob(IMethodReference methodRefere
                 return existing.Key;
             }
 
+            Debug.Assert(methodReference.CallingConvention != CallingConvention.Invalid);
```
This seems fine, but there don't appear to be any emit tests added, so I'm wondering how you ended up adding this?

333fred replied:
> This seems fine, but there don't appear to be any emit tests added, so I'm wondering how you ended up adding this?

I added a new option to the CallingConvention, and I wanted to verify we never attempted to emit something that was invalid. There's nothing that prompted me to add this other than that all current elements in CallingConvention are valid conventions.

---
In reply to: [359009433](https://github.com/dotnet/roslyn/pull/40431#discussion_r359009433) [](ancestors = 359009433)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -431,6 +432,9 @@ internal NamespaceOrTypeOrAliasSymbolWithAnnotations BindNamespaceOrTypeOrAliasS
                     return bindPointer();
 
+                case SyntaxKind.FunctionPointerType:
+                    return TypeWithAnnotations.Create(FunctionPointerTypeSymbol.CreateFunctionPointerTypeSymbolFromSource((FunctionPointerTypeSyntax)syntax, this, diagnostics, basesBeingResolved));
```
>TypeWithAnnotations.Create(FunctionPointerTypeSymbol.CreateFunctionPointerTypeSymbolFromSource((FunctionPointerTypeSyntax)syntax, this, diagnostics, basesBeingResolved)); [](start = 27, length = 170)

I think ```suppressUseSiteDiagnostics``` should be respected.

AlekseyTs replied:
I am not sure why we are being inconsistent about that in existing code as well. We either need to understand why it is Ok to be inconsistent or align all branches.

---
In reply to: [359051739](https://github.com/dotnet/roslyn/pull/40431#discussion_r359051739) [](ancestors = 359051739)

333fred replied:
> We either need to understand why it is Ok to be inconsistent or align all branches.

That's outside the scope of this PR, so I've filled an issue to track: [https://github.com/dotnet/roslyn/issues/40707](https://github.com/dotnet/roslyn/issues/40707)

---
In reply to: [359053343](https://github.com/dotnet/roslyn/pull/40431#discussion_r359053343) [](ancestors = 359053343,359051739)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+        {
+        }
+
+        public MethodSymbol Signature { get; private set; } = null!;
```
>public MethodSymbol Signature { get; private set; } = null!; [](start = 8, length = 60)

I think this property should be read-only. It should not be possible to update the value once set. Constructor should be able to do this without problems, I think.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+            convention switch
+            {
+                "" => CallingConvention.Default,
+                "cdecl" => CallingConvention.C,
```
>C [](start = 45, length = 1)

It looks like there is a Cdecl member

AlekseyTs replied:
At least there is one in System.​Runtime.​Interop​Services.CallingConvention. Perhaps we should add this alias into our type so that we could use language terminology throughout.

---
In reply to: [359063427](https://github.com/dotnet/roslyn/pull/40431#discussion_r359063427) [](ancestors = 359063427)

333fred commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+        public override bool IsReadOnly => false;
+        public override SymbolKind Kind => SymbolKind.FunctionPointer;
+        public override Symbol? ContainingSymbol => null;
+        public override ImmutableArray<Location> Locations => ImmutableArray<global::Microsoft.CodeAnalysis.Location>.Empty;
```
I think I actually fixed this bug the other day... 🙂

---
In reply to: [358991743](https://github.com/dotnet/roslyn/pull/40431#discussion_r358991743) [](ancestors = 358991743)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+        }
+
+
+        internal override void AddNullableTransforms(ArrayBuilder<byte> transforms)
```
>internal [](start = 8, length = 8)

It looks like there is an extra empty line above

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+
+        internal override bool GetUnificationUseSiteDiagnosticRecursive(ref DiagnosticInfo result, Symbol owner, ref HashSet<TypeSymbol> checkedTypes)
+        {
+            throw new NotImplementedException();
```
>throw new NotImplementedException(); [](start = 12, length = 36)

This should be implemented

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+
+        internal override ImmutableArray<NamedTypeSymbol> InterfacesNoUseSiteDiagnostics(ConsList<TypeSymbol>? basesBeingResolved = null)
+        {
+            throw new NotImplementedException();
```
>throw new NotImplementedException(); [](start = 12, length = 36)

This should be implemented, return an empty set

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+
+        internal override TypeSymbol MergeEquivalentTypes(TypeSymbol other, VarianceKind variance)
+        {
+            throw new NotImplementedException();
```
>throw new NotImplementedException(); [](start = 12, length = 36)

This should be implemented

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+
+        internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnotations, TypeWithAnnotations> transform)
+        {
+            throw new NotImplementedException();
```
>throw new NotImplementedException(); [](start = 12, length = 36)

This should be implemented

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+            throw new NotImplementedException();
+        }
+
+        // PROTOTYPE(func-ptr): Make source and metadata versions
```
>// PROTOTYPE(func-ptr): Make source and metadata versions [](start = 8, length = 57)

I don't believe this is needed. We don't have source and metadata version of arrays, pointers and constructed types

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal class FunctionPointerTypeSymbol : TypeSymbol
```
>FunctionPointerTypeSymbol [](start = 19, length = 25)

Both classes should override and implement GetUseSiteDiagnostics

AlekseyTs replied:
See what else is being overridden by PointerTypeSymbol, I would expect all of that overridden here too. Equality, GetHashCode, etc.

---
In reply to: [359072058](https://github.com/dotnet/roslyn/pull/40431#discussion_r359072058) [](ancestors = 359072058)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+        // PROTOTYPE(func-ptr): Make source and metadata versions
+        private sealed class FunctionPointerMethodSymbol : MethodSymbol
+        {
+            private readonly FunctionPointerTypeSymbol _parent;
```
>private readonly FunctionPointerTypeSymbol _parent; [](start = 12, length = 51)

It feels like ownership shouldn't matter. Similar as there is no ownership relationship between a PointerTypeSymbol and its underlying type. This is not a member.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+        private sealed class FunctionPointerMethodSymbol : MethodSymbol
+        {
+            private readonly FunctionPointerTypeSymbol _parent;
+            private ImmutableArray<ParameterSymbol> _lazyParameterSymbols = ImmutableArray<ParameterSymbol>.Empty;
```
>private ImmutableArray<ParameterSymbol> _lazyParameterSymbols = ImmutableArray<ParameterSymbol>.Empty; [](start = 12, length = 102)

I do not believe there is any value in parameters being lazy. There is no laziness for array, regular pointers and constructed types. This is similar, it is a type that is built around a bunch of other types, these are not members, not defined in separate metadata tables or statements in source

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+
+                if (syntax.Parameters.Count > 1)
+                {
+                    signature._lazyParameterSymbols = ParameterHelpers.MakeParameters(typeBinder,
```
>signature._lazyParameterSymbols [](start = 20, length = 31)

This field should be readonly and initialized in the constructor.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+            public override RefKind RefKind { get; }
+            public override TypeWithAnnotations ReturnTypeWithAnnotations { get; }
+            public override ImmutableArray<ParameterSymbol> Parameters => _lazyParameterSymbols;
+            // PROTOTYPE(func-ptr): Custom modifiers for ref readonly
```
>Custom modifiers for ref readonly [](start = 36, length = 33)

Custom modifiers should be supported in general

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+            public override TypeWithAnnotations ReturnTypeWithAnnotations { get; }
+            public override ImmutableArray<ParameterSymbol> Parameters => _lazyParameterSymbols;
+            // PROTOTYPE(func-ptr): Custom modifiers for ref readonly
+            public override ImmutableArray<CustomModifier> RefCustomModifiers => throw new NotImplementedException();
```
>throw new NotImplementedException(); [](start = 81, length = 36)

It doesn't feel like throwing is the right thing to do. An empty set would be a better temporary implementation.

333fred replied:
With the current tests I have, nothing hits this code. I'd rather leave it throwing an exception that will result in an obvious test failure later than make a silent input to a later test that affects code in a way I don't understand.

---
In reply to: [359078224](https://github.com/dotnet/roslyn/pull/40431#discussion_r359078224) [](ancestors = 359078224)

AlekseyTs replied:
> With the current tests I have, nothing hits this code. I'd rather leave it throwing an exception that will result in an obvious test failure later than make a silent input to a later test that affects code in a way I don't understand.

With either approach I think a PROTOTYPE comment is in order. Something should ensure we add tests for the code path.

---
In reply to: [360605306](https://github.com/dotnet/roslyn/pull/40431#discussion_r360605306) [](ancestors = 360605306,359078224)

333fred replied:
> With either approach I think a PROTOTYPE comment is in order.

That's why there's a prototype comment above it.

---
In reply to: [360612751](https://github.com/dotnet/roslyn/pull/40431#discussion_r360612751) [](ancestors = 360612751,360605306,359078224)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+            public override ImmutableArray<ParameterSymbol> Parameters => _lazyParameterSymbols;
+            // PROTOTYPE(func-ptr): Custom modifiers for ref readonly
+            public override ImmutableArray<CustomModifier> RefCustomModifiers => throw new NotImplementedException();
+            public override Symbol? ContainingSymbol => _parent;
```
>public override Symbol? ContainingSymbol => _parent; [](start = 12, length = 52)

I think this should return null.

AlekseyTs commented:
>                 var parameter = SourceParameterSymbol.Create(

We should not be creating SourceParameterSymbols for function pointers, they bring too much baggage that we don't want/need. We also would want to reuse the same type for function pointers imported from metadata or "retargeted".

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:105 in 81d5fc8. [](commit_id = 81d5fc825da51d3d2da3f3b64622dd73a53464c6, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6380,9 +6380,9 @@ private FunctionPointerTypeSyntax ParseFunctionPointerTypeSyntax()
+                var missingTypes = _pool.AllocateSeparated<ParameterSyntax>();
                 var missingTypeName = CreateMissingIdentifierName();
-                var missingType = SyntaxFactory.ModifiedType(default, missingTypeName);
+                var missingType = SyntaxFactory.Parameter(attributeLists: default, modifiers: default, missingTypeName, identifier: SyntaxFactory.Identifier(string.Empty), @default: null);
```
>Identifier [](start = 146, length = 10)

I think this should be missing identifier.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6408,7 +6408,7 @@ private FunctionPointerTypeSyntax ParseFunctionPointerTypeSyntax()
 
                         var parameterType = ParseTypeOrVoid();
-                        types.Add(SyntaxFactory.ModifiedType(modifiers, parameterType));
+                        types.Add(SyntaxFactory.Parameter(attributeLists: default, modifiers, parameterType, identifier: SyntaxFactory.Identifier(string.Empty), @default: null));
```
>SyntaxFactory.Identifier(string.Empty) [](start = 121, length = 38)

I think this should be missing identifier.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -0,0 +1,306 @@
+            Assert.Equal(SymbolKind.FunctionPointer, symbol.Kind);
+            Assert.Equal(TypeKind.FunctionPointer, symbol.TypeKind);
+            Assert.NotNull(symbol.Signature);
+            Assert.Equal(MethodKind.FunctionPointerSignature, symbol.Signature.MethodKind);
```
>Assert.Equal(MethodKind.FunctionPointerSignature, symbol.Signature.MethodKind); [](start = 12, length = 79)

Consider also asserting empty names for the method and all parameters

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+                    }
+                }
+
+                var signature = new FunctionPointerMethodSymbol(parent, callingConvention, refKind, returnType);
```
>callingConvention [](start = 72, length = 17)

Do we really need to reflect an invalid calling convention in the symbol? We reported an error, I think it would be fine to use CallingConvention.Default after that.

333fred replied:
> We reported an error, I think it would be fine to use CallingConvention.Default after that.

I considered that, but my thought was that this would eventually be exposed to consumers, so I wanted to explicitly indicate that we had a bad value.

---
In reply to: [359559897](https://github.com/dotnet/roslyn/pull/40431#discussion_r359559897) [](ancestors = 359559897)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -0,0 +1,306 @@
+        
+}", parseOptions: TestOptions.RegularPreview);
+
+            comp.VerifyDiagnostics(
```
>comp.VerifyDiagnostics( [](start = 12, length = 23)

Should also test what do symbols return in terms of modifiers related APIs

AlekseyTs commented:
It looks like we are missing support for generic type substitution and retargeting.

AlekseyTs commented:
Done with review pass (iteration 2)

333fred commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,264 @@
+        public override Accessibility DeclaredAccessibility => Accessibility.NotApplicable;
+        public override bool IsStatic => false;
+        public override bool IsAbstract => false;
+        public override bool IsSealed => false;
```
No, and not really, other than that imo `IsSealed` implies the presence of a `sealed` keyword, and none is here.

---
In reply to: [358992390](https://github.com/dotnet/roslyn/pull/40431#discussion_r358992390) [](ancestors = 358992390)

333fred commented:
I'll put a prototype comment in for these for the moment.

---
In reply to: [567208573](https://github.com/dotnet/roslyn/pull/40431#issuecomment-567208573) [](ancestors = 567208573)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [567208700](https://github.com/dotnet/roslyn/pull/40431#issuecomment-567208700) [](ancestors = 567208700)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_NameConflicts.cs:
```diff
@@ -25,11 +25,12 @@ private static Location GetLocation(Symbol symbol)
         }
 
-        internal void ValidateParameterNameConflicts(
+        internal void ValidateParameterNameConflicts<TParameterSymbol>(
```
><TParameterSymbol> [](start = 52, length = 18)

Do parameters in function pointers have names? Why would this helper be affected by the feature?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,136 @@
+            _signature = signature;
+        }
+
+        private FunctionPointerMethodSymbol _signature;
```
>private [](start = 8, length = 7)

readonly?

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -0,0 +1,611 @@
+            Assert.Equal(TypeKind.FunctionPointer, symbol.TypeKind);
+            Assert.NotNull(symbol.Signature);
+            Assert.Equal(MethodKind.FunctionPointerSignature, symbol.Signature.MethodKind);
+            Assert.Equal(string.Empty, symbol.Signature.Name);
```
>Assert.Equal(string.Empty, symbol.Signature.Name); [](start = 12, length = 50)

Consider also asserting other properties that have hardcoded values.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,136 @@
+
+        public override int GetHashCode()
+        {
+            return 19 * _signature.GetHashCode();
```
>19 [](start = 19, length = 2)

We usually do not use magic numbers to adjust hash codes. Do you expect the pointer and the underlying function be placed in the same dictionary? If it is really important to have distinct codes, could we use one of the functions we use to combine hash codes, or add one specifically for cases like this?

333fred replied:
> Do you expect the pointer and the underlying function be placed in the same dictionary?

They are both symbols, I wouldn't consider it out of the realm of possibility. I'll do something in our utils.

---
In reply to: [362668456](https://github.com/dotnet/roslyn/pull/40431#discussion_r362668456) [](ancestors = 362668456)

AlekseyTs replied:
> They are both symbols, I wouldn't consider it out of the realm of possibility. I'll do something in our utils.

Consider that hash code doesn't have to be unique (even unique ones often still end up in the same bucket), it just have to bee good enough to avoid major collisions. When two symbols collide, it is really not a problem I think. Especially, that I would find it extremely rare when someone would want to put both the pointer and its signature in the same dictionary.

---
In reply to: [362679025](https://github.com/dotnet/roslyn/pull/40431#discussion_r362679025) [](ancestors = 362679025,362668456)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,79 @@
+            Ordinal = ordinal;
+            _containingSymbol = containingSymbol;
+            RefCustomModifiers = refCustomModifiers;
+            Locations = locations;
```
>Locations = locations; [](start = 12, length = 22)

I don't think we need locations.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,79 @@
+        public override int Ordinal { get; }
+        public override Symbol ContainingSymbol => _containingSymbol;
+        public override ImmutableArray<CustomModifier> RefCustomModifiers { get; }
+        public override ImmutableArray<Location> Locations { get; }
```
>public override ImmutableArray<Location> Locations { get; } [](start = 8, length = 59)

This should always be empty, the same way as for containing ptr type and method

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,79 @@
+        public override Symbol ContainingSymbol => _containingSymbol;
+        public override ImmutableArray<CustomModifier> RefCustomModifiers { get; }
+        public override ImmutableArray<Location> Locations { get; }
+        public override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences => GetDeclaringSyntaxReferenceHelper<ParameterSyntax>(Locations);
```
>GetDeclaringSyntaxReferenceHelper<ParameterSyntax>(Locations); [](start = 85, length = 62)

Should always be empty

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,79 @@
+
+        public override int GetHashCode()
+        {
+            return Hash.Combine(_containingSymbol.GetHashCode(), GetHashCodeNoContaining());
```
>GetHashCodeNoContaining() [](start = 65, length = 25)

I think it is enough to use ```ordinal + 1```

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,79 @@
+
+        internal bool Equals(FunctionPointerParameterSymbol other, TypeCompareKind compareKind, IReadOnlyDictionary<TypeParameterSymbol, bool>? isValueTypeOverride)
+            => EqualsContainingVerified(other, compareKind, isValueTypeOverride)
+               && _containingSymbol.Equals(other._containingSymbol, compareKind, isValueTypeOverride);
```
> _containingSymbol.Equals(other._containingSymbol, compareKind, isValueTypeOverride); [](start = 17, length = 85)

Is this going to compare parameters? Is there a chance to go into a cycle? I think that it would be much simple if Equals.Override for the parameter symbol compared ordinal and containing FunctionPointerMethodSymbol. And the Equals implementation for FunctionPointerMethodSymbol compared parameters' properties other than ordinal, it would be fine to have a helper method for that in this class, just not called Equals.

333fred replied:
> Is there a chance to go into a cycle?

No, because the parent calls `EqualsContainingVerified`. That being said, I like the idea of the ordinal version, I'll change do that.

---
In reply to: [362685678](https://github.com/dotnet/roslyn/pull/40431#discussion_r362685678) [](ancestors = 362685678)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+    internal sealed class FunctionPointerMethodSymbol : MethodSymbol
+    {
+        private readonly ImmutableArray<FunctionPointerParameterSymbol> _parameters;
+        private DiagnosticInfo? _lazyUseSiteDiagnostic = CSDiagnosticInfo.EmptyErrorInfo;
```
>private DiagnosticInfo? _lazyUseSiteDiagnostic = CSDiagnosticInfo.EmptyErrorInfo; [](start = 8, length = 81)

I do not think we should be caching this for FunctionPointerMethodSymbol. We are caching this information only for definitions that do not belong to the current compilation. We also do not calculate that for definitions in source because those get all the required diagnostics while we bind syntax. Arrays, constructed types and regular pointer types are good examples. The symbols referring to the FunctionPointerMethodSymbol will merge information across all types involved and cache it as appropriate.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+        public override bool IsVirtual => false;
+        public override bool IsOverride => false;
+        public override bool IsAbstract => false;
+        public override bool IsSealed => true;
```
>true [](start = 41, length = 4)

false? It is not abstract, not virtual, not override. How could it be marked sealed?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+        internal override bool RequiresSecurityObject => false;
+        internal override bool IsDeclaredReadOnly => false;
+
+        internal override ImmutableArray<string> GetAppliedConditionalSymbols() => throw ExceptionUtilities.Unreachable;
```
>throw ExceptionUtilities.Unreachable [](start = 83, length = 36)

It feels like returning an empty set would be more appropriate.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+        internal override bool IsDeclaredReadOnly => false;
+
+        internal override ImmutableArray<string> GetAppliedConditionalSymbols() => throw ExceptionUtilities.Unreachable;
+        public override FlowAnalysisAnnotations ReturnTypeFlowAnalysisAnnotations => throw ExceptionUtilities.Unreachable;
```
>throw ExceptionUtilities.Unreachable [](start = 85, length = 36)

None?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+
+        internal override ImmutableArray<string> GetAppliedConditionalSymbols() => throw ExceptionUtilities.Unreachable;
+        public override FlowAnalysisAnnotations ReturnTypeFlowAnalysisAnnotations => throw ExceptionUtilities.Unreachable;
+        public override ImmutableHashSet<string> ReturnNotNullIfParameterNotNull => throw ExceptionUtilities.Unreachable;
```
>throw ExceptionUtilities.Unreachable [](start = 84, length = 36)

Empty?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+        internal override ImmutableArray<string> GetAppliedConditionalSymbols() => throw ExceptionUtilities.Unreachable;
+        public override FlowAnalysisAnnotations ReturnTypeFlowAnalysisAnnotations => throw ExceptionUtilities.Unreachable;
+        public override ImmutableHashSet<string> ReturnNotNullIfParameterNotNull => throw ExceptionUtilities.Unreachable;
+        public override FlowAnalysisAnnotations FlowAnalysisAnnotations => throw ExceptionUtilities.Unreachable;
```
>throw ExceptionUtilities.Unreachable [](start = 75, length = 36)

None?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+        internal override int CalculateLocalSyntaxOffset(int localPosition, SyntaxTree localTree) => throw ExceptionUtilities.Unreachable;
+        internal override IEnumerable<SecurityAttribute> GetSecurityInformation() => throw ExceptionUtilities.Unreachable;
+        internal override bool IsMetadataNewSlot(bool ignoreInterfaceImplementationChanges = false) => throw ExceptionUtilities.Unreachable;
+        internal override bool IsMetadataVirtual(bool ignoreInterfaceImplementationChanges = false) => throw ExceptionUtilities.Unreachable;
```
>throw ExceptionUtilities.Unreachable [](start = 103, length = 36)

False?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+        public override DllImportData GetDllImportData() => throw ExceptionUtilities.Unreachable;
+        internal override int CalculateLocalSyntaxOffset(int localPosition, SyntaxTree localTree) => throw ExceptionUtilities.Unreachable;
+        internal override IEnumerable<SecurityAttribute> GetSecurityInformation() => throw ExceptionUtilities.Unreachable;
+        internal override bool IsMetadataNewSlot(bool ignoreInterfaceImplementationChanges = false) => throw ExceptionUtilities.Unreachable;
```
>throw ExceptionUtilities.Unreachable [](start = 103, length = 36)

False?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed class FunctionPointerMethodSymbol : MethodSymbol
```
>FunctionPointerMethodSymbol [](start = 26, length = 27)

IsImplicitlyDeclared?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,79 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed class FunctionPointerParameterSymbol : ParameterSymbol
```
>FunctionPointerParameterSymbol [](start = 26, length = 30)

IsImplicitlyDeclared?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+
+        internal bool Equals(FunctionPointerMethodSymbol other, TypeCompareKind compareKind, IReadOnlyDictionary<TypeParameterSymbol, bool>? isValueTypeOverride)
+        {
+            return EqualsParametersVerified(other, compareKind, isValueTypeOverride)
```
>return EqualsParametersVerified(other, compareKind, isValueTypeOverride) [](start = 12, length = 72)

Consider optimizing for reference equality

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -21,16 +22,113 @@ internal static class ParameterHelpers
+                        refKind,
+                        syntax.Identifier,
+                        ordinal,
+                        (paramsKeyword.Kind() != SyntaxKind.None),
```
>(paramsKeyword.Kind() != SyntaxKind.None), [](start = 24, length = 42)

Consider using named argument.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -21,16 +22,113 @@ internal static class ParameterHelpers
+                        syntax.Identifier,
+                        ordinal,
+                        (paramsKeyword.Kind() != SyntaxKind.None),
+                        ordinal == 0 && thisKeyword.Kind() != SyntaxKind.None,
```
>ordinal == 0 && thisKeyword.Kind() != SyntaxKind.None [](start = 24, length = 53)

Consider using named argument.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -21,16 +22,113 @@ internal static class ParameterHelpers
+                {
+
+                    var locations = ImmutableArray.Create<Location>(new SourceLocation(syntax.Type));
+                    var modifierType = binder.GetWellKnownType(WellKnownType.System_Runtime_InteropServices_InAttribute, diagnostics, syntax);
```
>var modifierType = binder.GetWellKnownType(WellKnownType.System_Runtime_InteropServices_InAttribute, diagnostics, syntax); [](start = 20, length = 122)

Types and members should be looked up only when they are really needed. Otherwise their absence will break scenarios that don't need them.

AlekseyTs replied:
Consider extracting this logic and reusing across all kinds of parameter symbols built from syntax.

---
In reply to: [362938838](https://github.com/dotnet/roslyn/pull/40431#discussion_r362938838) [](ancestors = 362938838)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -21,16 +22,113 @@ internal static class ParameterHelpers
+                        ImmutableArray.Create(CSharpCustomModifier.CreateRequired(modifierType)) :
+                        ImmutableArray<CustomModifier>.Empty;
+
+                    if (parameterType.IsVoidType())
```
>if (parameterType.IsVoidType()) [](start = 20, length = 31)

Can we also share this logic with other parameters built from source?

333fred replied:
Other parameters report this at parsing time, because they don't have to know whether the current element is a return type that allows void, or a parameter type. Function pointers don't know whether it's a parameter until after parsing is done, so we have to do this at a later step.

---
In reply to: [362939946](https://github.com/dotnet/roslyn/pull/40431#discussion_r362939946) [](ancestors = 362939946)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,218 @@
+
+                returnType = typeBinder.BindType(returnTypeParameter.Type, diagnostics, basesBeingResolved, suppressUseSiteDiagnostics);
+
+                if (returnType.IsVoidType() && refKind != RefKind.None)
```
>if (returnType.IsVoidType() && refKind != RefKind.None) [](start = 16, length = 55)

Can we share this logic with other methods built from source?

333fred replied:
No, for similar reasons. Return types are parsed differently than parameter types in the main parser: if a `ref` is involved, they're parsed as a `RefTypeSyntax`. We can't reuse that in our parameter list because we need to support more than just `ref` and `ref readonly`, so we can't rely on the same paths. We _could_ potentially do this check while parsing, because having any modifiers on a `void` is never allowed, but I chose to do it here.

---
In reply to: [362940274](https://github.com/dotnet/roslyn/pull/40431#discussion_r362940274) [](ancestors = 362940274)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -21,16 +22,113 @@ internal static class ParameterHelpers
+                                        DiagnosticBag diagnostics) =>
+                {
+
+                    var locations = ImmutableArray.Create<Location>(new SourceLocation(syntax.Type));
```
>var locations = ImmutableArray.Create<Location>(new SourceLocation(syntax.Type)); [](start = 20, length = 81)

I think this shouldn't be needed for function pointers.

AlekseyTs commented:
>             binder.ValidateParameterNameConflicts(typeParameters, parameters, allowShadowingNames, diagnostics);

I don't think we should be calling this function for ```parsingFunctionPointer``` case. Even if today it is effectively do not have an observable effect, it is a bunch of  CPU cycles spent on a check that is not applicable to the scenario. 

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:209 in 974380e. [](commit_id = 974380e47a88c17ae5826729976fb9c6f79151c7, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -0,0 +1,611 @@
+                {
+                    Assert.False(ret1.Equals(ret2, TypeCompareKind.ConsiderEverything));
+                    Assert.False(ret1.Equals(ret2, TypeCompareKind.AllNullableIgnoreOptions));
+                    Assert.NotEqual(ret1.GetHashCode(), ret2.GetHashCode());
```
>Assert.NotEqual(ret1.GetHashCode(), ret2.GetHashCode()); [](start = 20, length = 56)

It is incorrect to test HashCode for inequality. There is no requirement for HashCode be different. There is a requirement for HashCode be equal for equal objects.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -0,0 +1,611 @@
+                {
+                    Assert.False(s1.Equals(s2, TypeCompareKind.ConsiderEverything));
+                    Assert.False(s1.Equals(s2, TypeCompareKind.AllNullableIgnoreOptions));
+                    Assert.NotEqual(s1.GetHashCode(), s2.GetHashCode());
```
>Assert.NotEqual(s1.GetHashCode(), s2.GetHashCode()); [](start = 20, length = 52)

Same here.

AlekseyTs commented:
Done with review pass (iteration 3)

AlekseyTs commented:
@333fred Do you know why CI jobs are failing?

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [570689355](https://github.com/dotnet/roslyn/pull/40431#issuecomment-570689355) [](ancestors = 570689355)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,82 @@
+
+        internal bool MethodEqualityChecks(FunctionPointerParameterSymbol other, TypeCompareKind compareKind, IReadOnlyDictionary<TypeParameterSymbol, bool>? isValueTypeOverride)
+            => RefKind == other.RefKind
+               && RefCustomModifiers.Equals(other.RefCustomModifiers)
```
>&& RefCustomModifiers.Equals(other.RefCustomModifiers) [](start = 15, length = 54)

Should modifiers be compared conditionally?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,82 @@
+        internal int MethodHashCode()
+            => Hash.Combine(TypeWithAnnotations.GetHashCode(),
+               Hash.Combine(RefKind.GetHashCode(),
+               Hash.Combine(Ordinal.GetHashCode(),
```
>Hash.Combine(Ordinal.GetHashCode(), [](start = 15, length = 35)

Is it really useful to include ordinal?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,82 @@
+            => Hash.Combine(TypeWithAnnotations.GetHashCode(),
+               Hash.Combine(RefKind.GetHashCode(),
+               Hash.Combine(Ordinal.GetHashCode(),
+               Hash.CombineValues(RefCustomModifiers))));
```
>Hash.CombineValues(RefCustomModifiers) [](start = 15, length = 38)

Is it the right thing to do to include custom modifiers into hash code?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -21,16 +22,110 @@ internal static class ParameterHelpers
+                                        DiagnosticBag diagnostics) =>
+                {
+
+                    var customModifiers = addRefReadOnlyModifier && refKind == RefKind.In ?
```
>addRefReadOnlyModifier && refKind == RefKind.In [](start = 42, length = 47)

I was really hoping that we'll move this condition into the ```CreateInModifiers``` helper. So that we could  completely get rid of the conditional logic here. ```SourceParameterSymbol``` could simply check if we have an empty set of modifiers in order to decide which class to instantiate.

333fred replied:
I don't particularly want to go that far in refactoring SourceParameterSymbol in this PR. I think having the check here is fine.

---
In reply to: [362993980](https://github.com/dotnet/roslyn/pull/40431#discussion_r362993980) [](ancestors = 362993980)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -21,16 +22,110 @@ internal static class ParameterHelpers
+                diagnostics,
+                allowRefOrOut: true,
+                allowThis: false,
+                addRefReadOnlyModifier: true,
```
>addRefReadOnlyModifier [](start = 16, length = 22)

When is this ```false``` for regular parameter case? It could be that the same situation can arise with function pointers as well. It is fine to add a PROTOTYPE comment to follow up later.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -107,7 +192,7 @@ internal static class ParameterHelpers
-            ImmutableArray<ParameterSymbol> parameters = builder.ToImmutableAndFree();
+            ImmutableArray<TParameterSymbol> parameters = builder.ToImmutableAndFree();
 
             var methodOwner = owner as MethodSymbol;
```
>var methodOwner = owner as MethodSymbol; [](start = 12, length = 40)

It feels like ```if (!parsingFunctionPointer)``` should be moved right above this line. All locals are calculated solely for the purpose of ValidateParameterNameConflicts call. The ```Debug.Assert(methodOwner?.MethodKind != MethodKind.LambdaMethod);``` will definitely be true for a function pointer, and, it feels like, the assert is tied to the following ```IDS_FeatureNameShadowingInNestedFunctions``` feature check anyway.

AlekseyTs commented on src/Workspaces/CoreTest/CodeStyle/NamingStylePreferencesUpgradeTests.cs:
```diff
@@ -20,15 +20,15 @@ private static void AssertTrimmedEqual(string expected, string actual)
         }
 
-        [Fact, Trait(Traits.Feature, Traits.Features.NamingStyle)]
+        [Fact(Skip = "PROTOTYPE(func-ptr): Adding a typekind is changing Struct to it's other value, Structure"), Trait(Traits.Feature, Traits.Features.NamingStyle)]
```
>"PROTOTYPE(func-ptr): Adding a typekind is changing Struct to it's other value, Structure" [](start = 21, length = 90)

The number ```FunctionPointer = 13,``` definitely has something to do with this.

333fred replied:
Yeah. Current guess is that the .NET serializer does a binary search among a sorted list of names, and it doesn't do anything special for duplicate values. Will need to work with the IDE as to whether they care.

---
In reply to: [362996793](https://github.com/dotnet/roslyn/pull/40431#discussion_r362996793) [](ancestors = 362996793)

AlekseyTs commented:
Done with review pass (iteration 6). It looks like remaining issues are mostly around testing/adjusting equality implementation in presence of custom modifiers.

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [570733478](https://github.com/dotnet/roslyn/pull/40431#issuecomment-570733478) [](ancestors = 570733478)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,81 @@
+        internal bool MethodEqualityChecks(FunctionPointerParameterSymbol other, TypeCompareKind compareKind, IReadOnlyDictionary<TypeParameterSymbol, bool>? isValueTypeOverride)
+            => RefKind == other.RefKind
+               && ((compareKind & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0
+                    ? RefCustomModifiers.Equals(other.RefCustomModifiers)
```
>Equals [](start = 41, length = 6)

Is this a value or a reference comparison?

AlekseyTs replied:
Do we have a test that relies on value equality of modifiers?

---
In reply to: [363539956](https://github.com/dotnet/roslyn/pull/40431#discussion_r363539956) [](ancestors = 363539956)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,81 @@
+
+        internal bool MethodEqualityChecks(FunctionPointerParameterSymbol other, TypeCompareKind compareKind, IReadOnlyDictionary<TypeParameterSymbol, bool>? isValueTypeOverride)
+            => RefKind == other.RefKind
+               && ((compareKind & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0
```
>((compareKind & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0 [](start = 18, length = 86)

I would expect some tests added around this condition.

333fred replied:
We're not reading in custom modifiers from metadata currently, so there's no real way to test this flag actually working. I'll add a prototype comment.

---
In reply to: [363541372](https://github.com/dotnet/roslyn/pull/40431#discussion_r363541372) [](ancestors = 363541372)

AlekseyTs commented:
Done with review pass (iteration 8)

RikkiGibson commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -0,0 +1,214 @@
+        public override ImmutableArray<Location> Locations => ImmutableArray<Location>.Empty;
+        public override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences => ImmutableArray<SyntaxReference>.Empty;
+        public override Accessibility DeclaredAccessibility => Accessibility.NotApplicable;
+        public override bool IsStatic => false;
```
I can't find an override of RequiresInstanceReceiver, doesn't this mean that other parts of the compiler may expect that calls to this function pointer require an instance receiver?

333fred replied:
This is just a method signature. The compiler won't be emitting calls to it, it'll be using `calli` with an appropriate calling convention.

RikkiGibson commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -0,0 +1,83 @@
+            => RefKind == other.RefKind
+               // PROTOTYPE(func-ptr): When we can read in arbitrary modifiers from metadata, test this flag
+               && ((compareKind & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0
+                    ? RefCustomModifiers.SequenceEqual(other.RefCustomModifiers)
```
nit: seems like there's trailing whitespace on line 53?

RikkiGibson commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -0,0 +1,762 @@
+
+                Assert.False(symbol.IsExtensionMethod);
+                Assert.False(symbol.HidesBaseMethodsByName);
+                Assert.False(symbol.IsStatic);
```
consider asserting on symbol.RequiresInstanceReceiver as well.

333fred replied:
I don't think asserting that is of any particular use. The method won't be used. If anything, I would almost consider making it an `ExceptionUtilities.Unreachable`

333fred commented:
Rebased on top of the updated feature branch.
